src/test/java/com/faforever/client/mapstruct/MapperSetup.java

```diff
@@ -0,0 +1,39 @@
+package com.faforever.client.mapstruct;
+
+import org.mapstruct.factory.Mappers;
+
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MapperSetup {
+  private static Map<Class<?>, Object> INJECTED_MAPPERS = new HashMap<>();
+
+  public static <T> void injectMappers(T mapper) throws IllegalAccessException {
+    Class<T> mapperClass = (Class<T>) mapper.getClass();
+    INJECTED_MAPPERS.put(mapperClass, mapper);
+    for (Field field : mapperClass.getDeclaredFields()) {
+      Class<?> internalMapperType = field.getType();
+      if (INJECTED_MAPPERS.containsKey(internalMapperType)) {
+        injectMapper(mapper, field, INJECTED_MAPPERS.get(internalMapperType));
+        continue;
+      }
+
+      Object internalMapper = Mappers.getMapper(internalMapperType);
+      if (internalMapper == null) {
+        continue;
+      }
+
+      injectMapper(mapper, field, internalMapper);
+
+      INJECTED_MAPPERS.put(internalMapperType, internalMapper);
+      injectMappers(internalMapper);
+    }
+  }
+
+  private static <T> void injectMapper(T mapper, Field field, Object internalMapper) throws IllegalAccessException {
+    field.setAccessible(true);
+    field.set(mapper, internalMapper);
+    field.setAccessible(false);
+  }
+}
```

Comment:
A stateful static class that injects dependencies in a Spring Application seems wrong. Why doesn't Spring's DI work?

Reply:
This is used for tests because spring injection does not work for service tests as the map service and mod service use a completable task which requires the javafx application to be running. However running the javafx application blocks the ability to run the tests in parallel.

Additionally none of the tests are currently configured to use the @SpringTest annotation which provides injection into tests. Using this would require a refactor of most of the tests.

Reply:
I am hoping to remove the dependencies of these services of the javafx UI and use proper spring tests but that is for another pr

Reply:
Ah I didn't realize it's in `test`.